crates/bevy_animation/src/curves.rs

```diff
@@ -0,0 +1,702 @@
+    /// Because 3 values are needed to perform cubic interpolation, `values` must have triple the
+    /// length of `times` — each consecutive triple `a_k, v_k, b_k` associated to time `t_k`
+    /// consists of:
+    /// - The in-tangent `a_k` for the sample at time `t_k`
+    /// - The actual value `v_k` for the sample at time `t_k`
+    /// - The out-tangent `b_k` for the sample at time `t_k`
+    ///
+    /// For example, for a curve built from two keyframes, the inputs would have the following form:
+    /// - `times`: `[t_0, t_1]`
+    /// - `values`: `[a_0, v_0, b_0, a_1, v_1, b_1]`
+    #[inline]
+    pub fn new(
+        times: impl IntoIterator<Item = f32>,
+        values: impl IntoIterator<Item = T>,
+    ) -> Result<Self, ChunkedUnevenCoreError> {
+        Ok(Self {
+            core: ChunkedUnevenCore::new(times, values, 3)?,
+        })
+    }
+}
+
+/// A curve specifying the translation component of a [`Transform`] in animation. The variants are
+/// broken down by interpolation mode (with the exception of `Constant`, which never interpolates).
+///
+/// This type is, itself, a `Curve<Vec3>`, and it internally uses the provided sampling modes; each
+/// variant "knows" its own interpolation mode.
+///
+/// [`Transform`]: bevy_transform::components::Transform
+
+#[derive(Clone, Debug, Reflect)]
+pub enum TranslationCurve {
+    /// A curve which takes a constant value over its domain. Notably, this is how animations with
+    /// only a single keyframe are interpreted.
+    Constant(ConstantCurve<Vec3>),
+
+    /// A curve which interpolates linearly between keyframes.
+    Linear(UnevenSampleAutoCurve<Vec3>),
+
+    /// A curve which interpolates between keyframes in steps.
+    Step(SteppedKeyframeCurve<Vec3>),
+
+    /// A curve which interpolates between keyframes by using auxiliary tangent data to join
+    /// adjacent keyframes with a cubic Hermite spline, which is then sampled.
+    CubicSpline(CubicKeyframeCurve<Vec3>),
+}
+
+impl Curve<Vec3> for TranslationCurve {
+    #[inline]
+    fn domain(&self) -> Interval {
+        match self {
+            TranslationCurve::Constant(c) => c.domain(),
+            TranslationCurve::Linear(c) => c.domain(),
+            TranslationCurve::Step(c) => c.domain(),
+            TranslationCurve::CubicSpline(c) => c.domain(),
+        }
+    }
+
+    #[inline]
+    fn sample_unchecked(&self, t: f32) -> Vec3 {
+        match self {
+            TranslationCurve::Constant(c) => c.sample_unchecked(t),
+            TranslationCurve::Linear(c) => c.sample_unchecked(t),
+            TranslationCurve::Step(c) => c.sample_unchecked(t),
+            TranslationCurve::CubicSpline(c) => c.sample_unchecked(t),
+        }
+    }
+}
+
+impl TranslationCurve {
+    /// The time of the last keyframe for this animation curve. If the curve is constant, None
+    /// is returned instead.
+    #[inline]
+    pub fn duration(&self) -> Option<f32> {
```

Comment:
I feel like the doc comment and the name of the function are not aligned. In my head `duration` is more like the length of the domain whereas the comment and the implementation are about the `end_time` or however you want to call it. If the curve implicitly will always start at `t = 0` then we should note that somewhere.

The same review comment applies to the other enums and their `duration` methods.

Reply:
I agree. I expect duration() to return a `std::time::duration` like `Time.delta()` would